test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go

```diff
@@ -0,0 +1,508 @@
+		cluster:           cluster,
+		machinePool:       mp,
+		dockerMachinePool: dmp,
+		logger:            log.WithValues("node-pool", dmp.Name),
+	}
+
+	if err := np.Refresh(); err != nil {
+		return np, errors.Wrapf(err, "failed to refresh the node pool")
+	}
+	return np, nil
+}
+
+func (np *nodePool) Refresh() error {
+	machinePoolLabels := map[string]string{"dmp": np.dockerMachinePool.Name}
+	machines, err := docker.ListMachinesByCluster(np.cluster.Name, machinePoolLabels, np.logger)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all machines in the cluster")
+	}
+
+	np.machines = machines
+	return nil
+}
+
+func (np *nodePool) MachinesMatchingSpec() []*docker.Machine {
+	var matchingMachines []*docker.Machine
+	for _, machine := range np.machines {
+		if !machine.IsControlPlane() && machine.ImageVersion() == *np.machinePool.Spec.Template.Spec.Version {
+			matchingMachines = append(matchingMachines, machine)
+		}
+	}
+
+	return matchingMachines
+}
+
+func (np *nodePool) MachinesNotMatchingSpec() []*docker.Machine {
+	var matchingMachines []*docker.Machine
+	for _, machine := range np.machines {
+		if !machine.IsControlPlane() && machine.ImageVersion() != *np.machinePool.Spec.Template.Spec.Version {
+			matchingMachines = append(matchingMachines, machine)
+		}
+	}
+
+	return matchingMachines
+}
+
+func (np *nodePool) AddNewMachine(ctx context.Context) error {
+	instanceName := util.RandomString(6)
+	externalMachine, err := docker.NewMachine(np.cluster.Name, instanceName, np.dockerMachinePool.Spec.Template.CustomImage, map[string]string{"dmp": np.dockerMachinePool.Name}, np.logger)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create helper for managing the externalMachine named %s", instanceName)
+	}
+
+	if err := externalMachine.Create(ctx, constants.WorkerNodeRoleValue, np.machinePool.Spec.Template.Spec.Version, np.dockerMachinePool.Spec.Template.ExtraMounts); err != nil {
+		return errors.Wrapf(err, "failed to create docker machine with instance name %s", instanceName)
+	}
+
+	np.dockerMachinePool.Status.Instances = append(np.dockerMachinePool.Status.Instances, &infrav1exp.DockerMachinePoolInstanceStatus{
+		InstanceName: instanceName,
+		Version:      np.machinePool.Spec.Template.Spec.Version,
+	})
+
+	return np.Refresh()
+}
+
+func (np *nodePool) ReconcileMachines(ctx context.Context) error {
+	matchingMachineCount := int32(len(np.MachinesMatchingSpec()))
+	if matchingMachineCount < *np.machinePool.Spec.Replicas {
+		for i := int32(0); i < *np.machinePool.Spec.Replicas-matchingMachineCount; i++ {
+			if err := np.AddNewMachine(ctx); err != nil {
+				return errors.Wrap(err, "failed to create a new docker machine")
+			}
+		}
+	}
+
+	for _, machine := range np.MachinesMatchingSpec() {
+		if err := np.reconcileMachine(ctx, machine); err != nil {
+			if errors.Is(err, &TransientError{}) {
+				return err
+			}
+			return errors.Wrap(err, "failed to reconcile machine")
+		}
+	}
+
+	return np.Refresh()
```

Comment:
should this goes before the loop reconciling machines, so the new instance are reconciled immediately?

Reply:
Refresh is called when initializing the node pool. If the node pool were to stick around for a while, this change would probably be needed. At this point, a node pool is created in the reconcile loop and is acting on only that one docker machine pool. I don't think that much will change between construction and use of the node pool, so this probably won't have much effect.